src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +409,115 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
             }
         }
 
+        private async Task<(CompletionList CompletionList, bool IsIncomplete, long ResultId)?> GetCompletionListAndFilterDownAsync(
+            LSP.CompletionParams request,
+            SourceText sourceText,
+            Document document,
+            OptionSet completionOptions,
+            CompletionService completionService,
+            CancellationToken cancellationToken)
+        {
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+            var completionTrigger = await ProtocolConversions.LSPToRoslynCompletionTriggerAsync(request.Context, document, position, cancellationToken).ConfigureAwait(false);
+
+            // Get the cached completion list in the case of a request for an incomplete list or calculate a new list and cache it.
+            CompletionList completionList;
+            long resultId;
+            if (request.Context?.TriggerKind == LSP.CompletionTriggerKind.TriggerForIncompleteCompletions)
+            {
+                // This is a request for a list we already calculated but was incomplete.
+                // Use the resultId to retrieve the completion list we cached for the original result.
+                Contract.ThrowIfFalse(_lastIncompleteResultId.HasValue, "Trigger for incomplete list missing previous result id");
+                var cachedList = _completionListCache.GetCachedCompletionList(_lastIncompleteResultId.Value);
+                Contract.ThrowIfNull(cachedList, "Trigger for incomplete list missing cached list");
+
+                completionList = cachedList.CompletionList;
+                resultId = _lastIncompleteResultId.Value;
+            }
+            else
+            {
+                // This is a new completion request, clear out the last result Id for incomplete results.
+                _lastIncompleteResultId = null;
+
+                completionList = await completionService.GetCompletionsAsync(document, position, completionTrigger, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
+                if (completionList == null || completionList.Items.IsEmpty || cancellationToken.IsCancellationRequested)
+                {
+                    return null;
+                }
+
+                // Cache the completion list so we can avoid recomputation in the resolve handler
+                resultId = _completionListCache.UpdateCache(request.TextDocument, completionList);
+            }
+
+            // Determine the current filter text based on the position in the document.
+            var filterText = sourceText.GetSubText(completionService.GetDefaultCompletionListSpan(sourceText, position)).ToString();
```

Comment:
I think it would be helpful for readability if the above and below parts were broken up into smaller functions (potentially local functions?), maybe something like `GetCompletionListAsync` and `ReduceCompletionList`?